ethereum/watcher/watcher.go

```diff
@@ -0,0 +1,100 @@
+package watcher
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	eth "github.com/ethereum/go-ethereum"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+var checkForBlocksTimeout = time.Second
+
+// EventFilterer filters the chain for specific events (logs).
+// When it finds a desired log, it puts it into its outbound channel.
+type EventFilterer struct {
+	ctx         context.Context
+	ec          *ethclient.Client
+	topic       ethcommon.Hash
+	filterQuery eth.FilterQuery
+	logCh       chan<- []ethtypes.Log
+}
+
+// NewEventFilterer returns a new *EventFilterer.
+func NewEventFilterer(
+	ctx context.Context,
+	ec *ethclient.Client,
+	contract ethcommon.Address,
+	fromBlock *big.Int,
+	topic ethcommon.Hash,
+	logCh chan<- []ethtypes.Log,
+) *EventFilterer {
+	filterQuery := eth.FilterQuery{
+		FromBlock: fromBlock,
+		Addresses: []ethcommon.Address{contract},
+	}
+
+	return &EventFilterer{
+		ctx:         ctx,
+		ec:          ec,
+		topic:       topic,
+		filterQuery: filterQuery,
+		logCh:       logCh,
+	}
+}
+
+// Start starts the EventFilterer. It watches the chain for logs.
```

Comment:
Can we add some comments somewhere about why we went with the implementation below?

The ethereum client has `SubscribeNewHead(...)` and `SubscribeFilterLogs(...)` which someone reading the code might expect, or might try to rewrite the code to use. I'm going to guess that polling is more reliable?

Reply:
yeah, I found geth's subscriptions really wonky and didn't really work, I can retry with `SubscribeFilterLogs` and see.